simulator/src/main/java/com/hedera/block/simulator/grpc/impl/ConsumerStreamGrpcClientImpl.java

```diff
@@ -0,0 +1,121 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.block.simulator.grpc.impl;
+
+import static com.hedera.block.simulator.metrics.SimulatorMetricTypes.Counter.LiveBlocksConsumed;
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.block.common.utils.Preconditions;
+import com.hedera.block.simulator.config.data.GrpcConfig;
+import com.hedera.block.simulator.grpc.ConsumerStreamGrpcClient;
+import com.hedera.block.simulator.metrics.MetricsService;
+import com.hedera.hapi.block.protoc.BlockStreamServiceGrpc;
+import com.hedera.hapi.block.protoc.SubscribeStreamRequest;
+import com.hedera.hapi.block.protoc.SubscribeStreamResponse;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.stub.StreamObserver;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import javax.inject.Inject;
+
+/**
+ * Implementation of {@link ConsumerStreamGrpcClient} that handles the consumption of blocks
+ * via gRPC streaming. This implementation manages the connection to the server and tracks
+ * metrics related to block consumption.
+ */
+public class ConsumerStreamGrpcClientImpl implements ConsumerStreamGrpcClient {
+    // Configuration
+    private final GrpcConfig grpcConfig;
+
+    // Service dependencies
+    private final MetricsService metricsService;
+
+    // gRPC components
+    private ManagedChannel channel;
+    private BlockStreamServiceGrpc.BlockStreamServiceStub stub;
+    private StreamObserver<SubscribeStreamResponse> consumerStreamObserver;
+
+    // State
+    private final List<String> lastKnownStatuses;
+
+    /**
+     * Constructs a new ConsumerStreamGrpcClientImpl with the specified configuration and metrics service.
+     *
+     * @param grpcConfig The configuration for gRPC connection settings
+     * @param metricsService The service for recording consumption metrics
+     * @throws NullPointerException if any parameter is null
+     */
+    @Inject
+    public ConsumerStreamGrpcClientImpl(
+            @NonNull final GrpcConfig grpcConfig, @NonNull final MetricsService metricsService) {
+        this.grpcConfig = requireNonNull(grpcConfig);
+        this.metricsService = requireNonNull(metricsService);
+        this.lastKnownStatuses = new ArrayList<>();
+    }
+
+    @Override
+    public void init() {
+        channel = ManagedChannelBuilder.forAddress(grpcConfig.serverAddress(), grpcConfig.port())
+                .usePlaintext()
+                .build();
+        stub = BlockStreamServiceGrpc.newStub(channel);
+        lastKnownStatuses.clear();
+    }
+
+    @Override
+    public void requestBlocks(long startBlock, long endBlock) throws InterruptedException {
+        Preconditions.requireWhole(startBlock);
+        Preconditions.requireWhole(endBlock);
```

Comment:
Maybe here would be a good idea to have a precondition check to verify that the `endBlock` >= `startBlock`, seems logical to me. If that is so, this is a very good candidate for a new `Preconditions` method, named for example `Preconditions#requireGreaterOrEqual(final long longToTest, final long longBase)` or something similar.

Reply:
Done 🤝